File: src/main/java/arcana/common/packets/PacketSender.java
package arcana.common.packets;

import arcana.Arcana;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.network.NetworkRegistry;
import net.minecraftforge.fml.network.simple.SimpleChannel;

public class PacketSender {
    private static final String PROTOCOL_VERSION = "1";
    public static final SimpleChannel INSTANCE = NetworkRegistry.newSimpleChannel(
        new ResourceLocation(Arcana.id, "main"),
        () -> PROTOCOL_VERSION,
        PROTOCOL_VERSION::equals,
        PROTOCOL_VERSION::equals
    );
    private static int msgId = 0;

    public static void init() {
        regPackets(new MarksPacket(), new ToggleMinigamePacket(), new DataSyncPacket());
    }



    private static void regPackets(PacketHandler<?> ... packets){
        for (PacketHandler<?> p : packets){
            regPacket(p);
        }
    }

    private static <T extends PacketHandler<?>> void regPacket(PacketHandler<T> packetInstance){
        INSTANCE.registerMessage(msgId++, (Class<T>) packetInstance.getClass(), packetInstance::encode, packetInstance::decode, packetInstance::handle);
    }
}

File: src/main/kotlin/arcana/common/packets/PacketSender.kt
package arcana.common.packets

import arcana.Arcana
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.network.NetworkRegistry

object PacketSender {
    private const val PROTOCOL_VERSION = "1"
    val INSTANCE = NetworkRegistry.newSimpleChannel(
        ResourceLocation(Arcana.id, "main"),
        { PROTOCOL_VERSION },
        PROTOCOL_VERSION::equals,
        PROTOCOL_VERSION::equals)
    private var msgId = 0

    fun init() {
        regPackets(
            MarksPacket(),
            ToggleMinigamePacket(),
            DataSyncPacket(),
            ManaPacket(),
            KnowledgePacket()
        )
    }

    private fun regPackets(vararg packets: PacketHandler<*>) {
        for (p in packets) {
            regPacket(p)
        }
    }

    private fun <T : PacketHandler<T>> regPacket(packetInstance: PacketHandler<T>) {
        INSTANCE.registerMessage(msgId++, packetInstance.javaClass as Class<T>, packetInstance::encode, packetInstance::decode, packetInstance::handle)
    }
}